Implemented relaxed_swizzle SIMD opcode.

danlliu opened the pull request:
#### e1ec14877b7db64cf484cbd62480417029127c05
<pre>
Implemented relaxed_swizzle SIMD opcode.
<a href="https://bugs.webkit.org/show_bug.cgi?id=257878">https://bugs.webkit.org/show_bug.cgi?id=257878</a>
rdar://problem/110495643

Reviewed by Justin Michaud.

Added support for relaxed_swizzle as part of implementing relaxed SIMD functionality.

* JSTests/wasm/stress/simd-const-swizzle.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.i8x16.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.v128.const.i8x16.0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.i8x16.relaxed_swizzle.return.i8x16.extract_lane_u.1.async test):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDV_VV):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDV_VV):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):
* JSTests/wasm/libwabt.js:
(WabtModule.):
(WabtModule):
(set get if): Deleted.
* JSTests/wasm/stress/simd-const-relaxed-swizzle.js: Renamed from JSTests/wasm/stress/simd-const-swizzle.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.i8x16.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.v128.const.i8x16.0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.i8x16.relaxed_swizzle.return.i8x16.extract_lane_u.1.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.i8x16.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.v128.const.i8x16.0.8.2.3.4.5.6.7.8.9.10.11.12.13.14.15.i8x16.relaxed_swizzle.return.i8x16.extract_lane_u.1.async test):
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:

Canonical link: <a href="https://commits.webkit.org/265075@main">https://commits.webkit.org/265075@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60953701bdc30fbc513b0b11bf64a908f5fbd3a4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9611 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9868 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10112 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11273 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9397 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11847 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9823 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12323 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9763 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10622 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8184 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/11432 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7917 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8740 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16150 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/8183 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9012 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8889 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12227 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/9150 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9376 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7618 "5 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/9767 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8567 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2405 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2336 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12788 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/10024 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9150 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2467 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/b3/B3SIMDValue.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "b3/B3Opcode.h"
```
nit:  add -header-insertion=never to your server arguments to stop it from inserting silly headers all the time

justinmichaud commented on Source/JavaScriptCore/b3/B3Value.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "B3Value.h"
+#include "b3/B3Opcode.h"
```
nit

justinmichaud commented on Source/JavaScriptCore/b3/B3Value.cpp:
```diff
@@ -992,6 +994,7 @@ ValueKey Value::key() const
         numChildrenForKind(kind(), 3);
         return ValueKey(kind(), type(), as<SIMDValue>()->simdInfo(), child(0), child(1), child(2));
     case VectorSwizzle:
+    case VectorRelaxedSwizzle:
```
we should ensure only two children

justinmichaud commented on Source/JavaScriptCore/b3/B3Value.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "b3/B3Opcode.h"
```
nit

justinmichaud commented on Source/JavaScriptCore/b3/B3Value.h:
```diff
@@ -553,6 +554,7 @@ class JS_EXPORT_PRIVATE Value {
         case CheckMul:
         case Patchpoint:
         case VectorSwizzle:
+        case VectorRelaxedSwizzle:
```
2 children

justinmichaud commented on Source/JavaScriptCore/b3/B3ValueInlines.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "b3/B3Opcode.h"
```
nit

justinmichaud commented on Source/JavaScriptCore/b3/B3ValueKey.cpp:
```diff
@@ -25,6 +25,7 @@
 
 #include "config.h"
 #include "B3ValueKey.h"
+#include "b3/B3Opcode.h"
```
nit

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -25,6 +25,8 @@
 
 #pragma once
 
+#include "runtime/Options.h"
```
nit

justinmichaud reviewed (commented):
LGTM with the requested chagnes

justinmichaud commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -25,6 +25,7 @@
 
 #pragma once
 
+#include "wasm/WasmSIMDOpcodes.h"
```
nit

justinmichaud commented on Source/JavaScriptCore/b3/B3Validate.cpp:
```diff
@@ -670,6 +670,18 @@ class Validater {
+            // Relaxed SIMD
+            case VectorRelaxedSwizzle:
+                VALIDATE(!value->kind().hasExtraBits(), ("At ", *value));
+                VALIDATE(value->numChildren() == 2 || value->numChildren() == 3, ("At ", *value));
```
2 children

justinmichaud commented on Source/JavaScriptCore/b3/B3Value.cpp:
```diff
@@ -995,6 +996,10 @@ ValueKey Value::key() const
         if (numChildren() == 2)
             return ValueKey(kind(), type(), as<SIMDValue>()->simdInfo(), child(0), child(1), nullptr);
         return ValueKey(kind(), type(), as<SIMDValue>()->simdInfo(), child(0), child(1), child(2));
+    case VectorRelaxedSwizzle:
```
Let's add this to the 2 child case above, which should handle asserting that the number of children is always 2

justinmichaud reviewed (approved):
r=me again

danlliu committed:
Implemented relaxed_swizzle SIMD opcode.
https://bugs.webkit.org/show_bug.cgi?id=257878
rdar://problem/110495643

Reviewed by Justin Michaud.

Added support for relaxed_swizzle as part of implementing relaxed SIMD functionality.

* JSTests/wasm/stress/simd-const-swizzle.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.i8x16.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.v128.const.i8x16.0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.i8x16.relaxed_swizzle.return.i8x16.extract_lane_u.1.async test):
* Source/JavaScriptCore/b3/B3LowerToAir.cpp:
* Source/JavaScriptCore/b3/B3Opcode.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/b3/B3Opcode.h:
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Validate.cpp:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::effects const):
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
(JSC::B3::ValueKey::materialize const):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::addSIMDV_VV):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::addSIMDV_VV):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::simd):
* Source/JavaScriptCore/wasm/WasmSIMDOpcodes.h:
(JSC::dumpSIMDLaneOperation):
(JSC::isRelaxedSIMDOperation):
* JSTests/wasm/libwabt.js:
(WabtModule.):
(WabtModule):
(set get if): Deleted.
* JSTests/wasm/stress/simd-const-relaxed-swizzle.js: Renamed from JSTests/wasm/stress/simd-const-swizzle.js.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.i8x16.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.v128.const.i8x16.0.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.i8x16.relaxed_swizzle.return.i8x16.extract_lane_u.1.func.export.string_appeared_here.param.sz.i32.result.i32.i32.const.9.drop.v128.const.i8x16.1.2.3.4.5.6.7.8.9.10.11.12.13.14.15.16.v128.const.i8x16.0.8.2.3.4.5.6.7.8.9.10.11.12.13.14.15.i8x16.relaxed_swizzle.return.i8x16.extract_lane_u.1.async test):
* Source/JavaScriptCore/b3/B3SIMDValue.h:
* Source/JavaScriptCore/b3/B3Value.cpp:
(JSC::B3::Value::key const):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/b3/B3ValueInlines.h:
* Source/JavaScriptCore/b3/B3ValueKey.cpp:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:

Canonical link: https://commits.webkit.org/265075@main
